File: JS-CS-Detection-byExample/Dataset (ALERT 5 GB)/374130/spot-0.8/spot-0.8/src/evtgba/symbol.cc
// Copyright (C) 2004  Laboratoire d'Informatique de Paris 6 (LIP6),
// département Systèmes Répartis Coopératifs (SRC), Université Pierre
// et Marie Curie.
//
// This file is part of Spot, a model checking library.
//
// Spot is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// Spot is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
// License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Spot; see the file COPYING.  If not, write to the Free
// Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.

#include <cassert>
#include "symbol.hh"
#include <ostream>

namespace spot
{
  symbol::map symbol::instances_;

  symbol::symbol(const std::string* name)
    : name_(name), refs_(1)
  {
  }

  symbol::~symbol()
  {
    map::iterator i = instances_.find(name());
    assert (i != instances_.end());
    instances_.erase(i);
  }

  const symbol*
  symbol::instance(const std::string& name)
  {
    map::iterator i = instances_.find(name);
    if (i != instances_.end())
      {
	const symbol* s = i->second;
	s->ref();
	return s;
      }
    // Convoluted insertion because we want the NAME member of the
    // value to point to the key.
    i = instances_.insert(map::value_type(name, 0)).first;
    i->second = new symbol(&i->first);
    return i->second;
  }

  const std::string&
  symbol::name() const
  {
    return *name_;
  }

  void
  symbol::ref() const
  {
    ++refs_;
  }

  void
  symbol::unref() const
  {
    assert(refs_ > 0);
    --refs_;
    if (!refs_)
      delete this;
  }

  int
  symbol::ref_count_() const
  {
    return refs_;
  }

  unsigned
  symbol::instance_count()
  {
    return instances_.size();
  }

  std::ostream&
  symbol::dump_instances(std::ostream& os)
  {
    for (map::iterator i = instances_.begin(); i != instances_.end(); ++i)
      {
	os << i->second << " = " << i->second->ref_count_()
	   << " * symbol(" << i->second->name() << ")" << std::endl;
      }
    return os;
  }







}

File: JS-CS-Detection-byExample/Dataset (ALERT 5 GB)/911446/Alpha 0.1/quote/Symbol.java
/* Merchant of Venice - technical analysis software for the stock market.
   Copyright (C) 2002 Andrew Leppard (aleppard@picknowl.com.au)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA 
*/

package org.mov.quote;

import java.util.*;

public class Symbol implements Comparable {
    
    // Pack down symbol string, e.g. "ABCDE" into single int
    private int symbol;
    
    // Enough room to store "ABCDE". This value can be 1->6.
    public final static int MAXIMUM_SYMBOL_LENGTH = 5;
    
    // A-Z contains 26 letters which fit in 5 bits
    private final static int BITS_PER_CHARACTER = 5;

    public Symbol() {
	symbol = 0;
    }
    
    public Symbol(String symbolString) {
        setSymbol(symbolString);
    }
    
    public String getSymbol() {
        if(symbol == 0)
            return null;
	
        String symbolString = new String();
        int i = 0;
        boolean isMoreCharacters = true;
	char[] characters = new char[MAXIMUM_SYMBOL_LENGTH];
	
        while(isMoreCharacters) {
	    
            // 1..26 (A..Z)
            int characterNumber = symbol >> (BITS_PER_CHARACTER * i);
            characterNumber &= ((2 << BITS_PER_CHARACTER - 1) - 1);
	    
            if(characterNumber == 0)
                isMoreCharacters = false;
            else 
		characters[i++] = (char)(characterNumber - 1 + (int)'A');
        }
	
	return String.copyValueOf(characters, 0, i);
    }
    
    public void setSymbol(String symbolString) {
	
        symbolString = symbolString.toUpperCase();
	
        symbol = 0;
	
        assert symbolString.length() <= MAXIMUM_SYMBOL_LENGTH;

        for(int i = 0; i < symbolString.length(); i++) {

	    // 1..26 (A..Z)
	    int characterNumber = ((int)symbolString.charAt(i) + 1 - 
				   (int)'A');
	    
	    assert characterNumber <= BITS_PER_CHARACTER * 8;
	    symbol += characterNumber << (BITS_PER_CHARACTER * i);
	}
    }
    
    public int compareTo(Object object) {
	return toString().compareTo(object.toString());
    }
    
    public boolean equals(Object object) {
	return object.hashCode() == hashCode();
    }
    
    public int hashCode() {
        return symbol;
    }
    
    public String toString() {
        return getSymbol();
    }
}

